pipeline/foraging_model.py

```diff
@@ -0,0 +1,301 @@
+    prediction_accuracy: float    # non-cross-validated prediction accuracy 
+    cross_valid_accuracy_fit: float     # cross-validated accuracy (fitting set)
+    cross_valid_accuracy_test: float    # cross-validated accuracy (testing set)
+    cross_valid_accuracy_test_bias_only = NULL: float    # accuracy predicted only by bias (testing set)
+    """
+
+    key_source = (foraging_analysis.SessionTaskProtocol() & 'session_task_protocol = 100' & 'session_real_foraging'
+                  ) * Model() #& (experiment.Session & 'session_date > "2021-01-01"')
+
+    class FittedParam(dj.Part):
+        definition = """
+        -> master
+        -> Model.Param
+        ---
+        fitted_value: float
+        """
+    
+    class PredictiveChoiceProb(dj.Part):
+        definition = """
+        # Could be used to compute latent value Q (for models that have Q). Ignored trial skipped.
+        -> master
+        -> experiment.SessionTrial
+        -> experiment.WaterPort
+        ---
+        choice_prob: float        
+        """
+
+    def make(self, key):
+        choice_history, reward_history, p_reward, q_choice_outcome = get_session_history(key)
+        model_str = (Model & key).fetch('fit_cmd')
+
+        # --- Actual fitting ---
+        model_comparison_this = BanditModelComparison(choice_history, reward_history, model=model_str)
+        model_comparison_this.fit(pool='', plot_predictive=None, if_verbose=False)  # Parallel on sessions, not on DE
+        model_comparison_this.cross_validate(pool='', k_fold=2, if_verbose=False)
+
+        # ------ Grab results ----
+        fit_result = model_comparison_this.results_raw[0]
+        cross_valid_result = model_comparison_this.prediction_accuracy_CV.iloc[0]
+
+        # Insert session fitted stats
+        self.insert1(dict(**key,
+                          n_trials=fit_result.n_trials,
+                          n_params=fit_result.k_model,
+                          log_likelihood=fit_result.log_likelihood,
+                          aic=fit_result.AIC,
+                          bic=fit_result.BIC,
+                          lpt=fit_result.LPT,
+                          lpt_aic=fit_result.LPT_AIC,
+                          lpt_bic=fit_result.LPT_BIC,
+                          prediction_accuracy=fit_result.prediction_accuracy,
+                          cross_valid_accuracy_fit=np.mean(cross_valid_result.prediction_accuracy_fit),
+                          cross_valid_accuracy_test=np.mean(cross_valid_result.prediction_accuracy_test),
+                          cross_valid_accuracy_test_bias_only=np.mean(cross_valid_result.prediction_accuracy_test_bias_only),
+                          ),
+                     skip_duplicates=True, allow_direct_insert=True)
+
+        # Insert fitted params
+        for param, x in zip((Model.Param & key).fetch('model_param'), fit_result.x):
+            self.FittedParam.insert1(dict(**key, model_param=param, fitted_value=x))
```

Comment:
```suggestion
        self.FittedParam.insert([dict(**key, model_param=param, fitted_value=x) 
                         for param, x in zip((Model.Param & key).fetch('model_param'), fit_result.x)])
```
I'd suggest using batch insert with list comprehension here, boosting the insertion performance.

Though this relies on the fact that `(Model.Param & key).fetch('model_param')` returns the list of `model_param` in the same order as `fit_result.x`, is this always guaranteed?

Reply:
Yes, the list comprehension is better, but the order is a good catch!! 

In fact, I just checked the order and they are NOT aligned... The param order of `fit_result.x` follows the order specified in `MODELS` of `Model.load_models()`. Although I used the same order to insert `Model.Param`, the params were not indexed, so the order is not kept during `fetch`. And I haven't looked into `FittedParam` after fitting my test sessions so I didn't realize this bug. Thank you for pointing this out!

To fix this, I add a `param_idx` attribute to `Model.Param` and an `order_by` to `fetch`. Is this the best way?

Reply:
Given my understanding that there's no other way of knowing the param ordering returned by `fit_result`, then yeah, adding another attribute `param_idx` is necessary. I was hoping to avoid having to introduce another attribute, e.g. if `fit_result` is returned as a dictionary, but not the case, so this is probably best.